sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/ThrottleWithoutExternalResource.java

```diff
@@ -17,41 +17,651 @@
+      return new OffsetRangeTracker(restriction);
+    }
+
+    /** Simply returns the {@link OffsetRange} restriction. */
+    @TruncateRestriction
+    public RestrictionTracker.TruncateResult<OffsetRange> truncate(
+        @Restriction OffsetRange restriction) {
+      return new RestrictionTracker.TruncateResult<OffsetRange>() {
+        @Override
+        public ThrottleWithoutExternalResource.@Nullable OffsetRange getTruncatedRestriction() {
+          return restriction;
+        }
+      };
+    }
+
+    /**
+     * The {@link GetInitialWatermarkEstimatorState} initializes to this DoFn's output watermark to
+     * a negative infinity timestamp via {@link BoundedWindow#TIMESTAMP_MIN_VALUE}. The {@link
+     * Instant} returned by this method provides the runner the value is passes as an argument to
+     * this DoFn's {@link #newWatermarkEstimator}.
+     */
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkState() {
+      return BoundedWindow.TIMESTAMP_MIN_VALUE;
+    }
+
+    /**
+     * This DoFn uses a {@link WatermarkEstimators.Manual} as its {@link NewWatermarkEstimator},
+     * instantiated from an {@link Instant}. The state argument in this method comes from the return
+     * of the {@link #getInitialWatermarkState}.
+     */
+    @NewWatermarkEstimator
+    public WatermarkEstimator<Instant> newWatermarkEstimator(
+        @WatermarkEstimatorState Instant state) {
+      return new WatermarkEstimators.Manual(state);
+    }
+
+    @ProcessElement
+    public void process(
+        @Element KV<Integer, List<RequestT>> element,
+        ManualWatermarkEstimator<Instant> estimator,
+        RestrictionTracker<OffsetRange, Integer> tracker,
+        MultiOutputReceiver receiver) {
+
+      int size = 0;
+      if (element.getValue() != null) {
+        size = element.getValue().size();
+      }
+
+      incIfPresent(inputElementsCounter, size);
+
+      if (element.getValue() == null || element.getValue().isEmpty()) {
+        return;
+      }
+
+      while (tracker.tryClaim(tracker.currentRestriction().getCurrent() + 1)) {
+        Instant nextEmittedTimestamp =
+            estimator.currentWatermark().plus(configuration.getMaximumRate().getInterval());
+        int index = tracker.currentRestriction().getCurrent();
+        RequestT requestT = element.getValue().get(index);
+        outputAndSetWatermark(nextEmittedTimestamp, requestT, estimator, receiver);
+      }
+    }
+
+    /**
+     * Emits the element at the nextEmittedTimestamp and sets the watermark to the same when {@link
+     * Instant#now()} reaches the nextEmittedTimestamp.
+     */
+    private void outputAndSetWatermark(
+        Instant nextEmittedTimestamp,
+        RequestT requestT,
+        ManualWatermarkEstimator<Instant> estimator,
+        MultiOutputReceiver receiver) {
+      Instant now = Instant.now();
+      while (now.isBefore(nextEmittedTimestamp)) {
+        now = Instant.now();
```

Comment:
I think we maybe discussed this before, but I can't remember; is there a reason we can't use process continuations to handle this instead of live looping? I'm worried about this wasting compute and potentially leading to runners like Dataflow scaling up.

Reply:
https://beam.apache.org/documentation/programming-guide/#user-initiated-checkpoint

Reply:
The reason I am not returning a ProcessContinuation is ["While the runner tries to honor the resume time, this is not guaranteed"](https://beam.apache.org/documentation/programming-guide/#user-initiated-checkpoint:~:text=While%20the%20runner%20tries%20to%20honor%20the%20resume%20time%2C%20this%20is%20not%20guaranteed.). The main feature of this implementation is to control the time when elements are emitted and not give up control.